### DISCUSSION TITLE
SSLHandshakeException: PKIX path building failed:

### DISCUSSION THREAD

**Discussion by iamuspace** [langchain4j/langchain4j#347] [General]
OpenAiChatModel.generate() error. 
stand alone java application code works fine. 

but, 
error in web application using Glassfish web server of NetBeans IDE

I guess that SSL to api.openai.com cert. 
I added api.openai.com's cert to Zulu-21 OpenJDK  security/cacerts using keytool. 

The same error still occurs.

How can I solve the problem?

java.lang.RuntimeException: java.lang.RuntimeException: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at dev.langchain4j.internal.RetryUtils.withRetry(RetryUtils.java:29)
	at dev.langchain4j.model.openai.OpenAiChatModel.generate(OpenAiChatModel.java:127)
	at dev.langchain4j.model.openai.OpenAiChatModel.generate(OpenAiChatModel.java:91)

**Comment by iamuspace** [comment]
I worked fine with Tomcat and NetBeans. 

Glassfish error. 

I don't know for Glassfish error for cert.

I recommand Tomcat with NetBeans for LangChain4j.

**Comment by jawaharbabus** [comment]
Hello, what is the status? I am using Glassfish 7 and I am facing this error. 
on an unrelated note, I am facing a few other issues with langchain4j-openai with Jakarta ee which is not the case with a standalone application. I finally planned to switch to Ollama which seems to work fine in my case.
I used TomEE server and it solved the issue.

**Reply by jawaharbabus** [reply]
Sure, I will take a deeper look at this issue.

**Comment by MeDeity** [comment]
I also encountered this problem and used openssl to check that the certificate chain has been successfully added to the JDK certificate repository, but it still reports PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target, Can we report OKHTTPCLIENT for developers to customize, so that they can make decisions on whether to skip this certificate verification, for example
In package dev.langchain4j.community.model.zhipu.ZhipuAiClient, we have opened okFHIR for users to customize
```fake code
public getHttpClient():any{
return this.okHttpClient;
}
```

**Comment by feodnkn** [comment]
Here is what I've done for gemini, I think it should help with open AI

/** proxy settings if you use proxy */
System.setProperty("http.proxyHost", "XXX");
System.setProperty("http.proxyPort", "XXX");
System.setProperty("java.net.useSystemProxies", "true");

/**
In case you experiece an SSL Handshake error, you might have to import the certificate manually into the keystore and set the path as a system property
*/

System.setProperty("javax.net.ssl.trustStore", "[Path to JDK 11]\\lib\\security\\cacerts");
System.setProperty("javax.net.ssl.trustStorePassword", "XXX");
